### PR TITLE
Fix database  does not exists error when is_member(), schema_id() function gets called in parallel query mode.

### PR DESCRIPTION
**Pull request by ahmed-shameem**
### Description

This PR intends to solve the issue of parallel worker not being able to obtain correct logical database name. Whenever get_cur_db_name() is called, it returns static char current_db_name[] which is initialised as empty. Now, parallel workers are not aware of the database name. For them, it is empty whenever get_cur_db_name() gets called. Hence, we were getting the following error: 
```
database "" does not exist
```
So, to communicate the logical database name and more data related to babelfish we have introduced a struct BabelfishFixedParallelState. Currently BabelfishFixedParallelState has only 1 field, i.e., logical_db_name, we can add more fields to it whenever necessary. Then we will write the logical_db_name in DSM during initialisation (in InitializeParallelDSM) using a hook. Another hook is used while trying to fetch the logical database name in ParallelWorkerMain().

In this way we are introduction a mechanism through which we can communicate any babelfish related information to parallel workers.

Signed-off-by: Shameem Ahmed [shmeeh@amazon.com](mailto:rcv@amazon.com)
 
### Issues Resolved

BABEL-4538, BABEL-4481, BABEL-4421
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/access/transam/parallel.c [hunk 1]
```diff
@@ -1486,6 +1501,12 @@ ParallelWorkerMain(Datum main_arg)
 	/* Attach to the leader's serializable transaction, if SERIALIZABLE. */
 	AttachSerializableXact(fps->serializable_xact_handle);
 
+
+	/* Hook for babelfish to restore babelfish fixed parallel state */
+	if (MyFixedParallelState->babelfish_context && babelfixedparallelstate_restore_hook)
```
Comment (Deepesh125) [inline src/backend/access/transam/parallel.c]: Move this above `RestoreUncommittedEnums(uncommittedenumsspace);`

File: src/backend/access/transam/parallel.c [hunk 2]
```diff
@@ -77,6 +77,10 @@
 #define PARALLEL_KEY_RELMAPPER_STATE		UINT64CONST(0xFFFFFFFFFFFF000D)
 #define PARALLEL_KEY_UNCOMMITTEDENUMS		UINT64CONST(0xFFFFFFFFFFFF000E)
 
+/* Hooks for communicating babelfish related information to parallel worker */
+babelfixedparallelstate_insert_hook_type babelfixedparallelstate_insert_hook;
+babelfixedparallelstate_restore_hook_type babelfixedparallelstate_restore_hook;
+
```
Comment (Deepesh125) [inline src/backend/access/transam/parallel.c]: Hook name should be generic, does not need to have babelfish in the name. Also it should be based on the original function ie, InitializeParallelDSM.

File: src/backend/access/transam/parallel.c [hunk 3]
```diff
@@ -291,8 +295,14 @@ InitializeParallelDSM(ParallelContext *pcxt)
...
 							   strlen(pcxt->function_name) + 2);
 		shm_toc_estimate_keys(&pcxt->estimator, 1);
+
+		/* Estimate how much we'll need for the babelfish fixed parallel state */
+		if (MyProcPort->is_tds_conn && babelfixedparallelstate_insert_hook)
+			(*babelfixedparallelstate_insert_hook) (pcxt, true);
 	}
 
+	
+
```
Comment (Deepesh125) [inline src/backend/access/transam/parallel.c]: Unnecessary changes

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: src/backend/access/transam/parallel.c (hunk 1 above)
Comment (ahmed-shameem) [reply src/backend/access/transam/parallel.c]: Updated in latest rev.

File: src/backend/access/transam/parallel.c (hunk 2 above)
Comment (ahmed-shameem) [reply src/backend/access/transam/parallel.c]: Updated in latest rev.

File: src/backend/access/transam/parallel.c (hunk 3 above)
Comment (ahmed-shameem) [reply src/backend/access/transam/parallel.c]: Updated in latest rev.

File: src/backend/access/transam/parallel.c [hunk 4]
```diff
@@ -77,6 +77,10 @@
 #define PARALLEL_KEY_RELMAPPER_STATE		UINT64CONST(0xFFFFFFFFFFFF000D)
 #define PARALLEL_KEY_UNCOMMITTEDENUMS		UINT64CONST(0xFFFFFFFFFFFF000E)
 
+/* Hooks for communicating babelfish related information to parallel worker */
+InitializeParallelDSM_hook_type InitializeParallelDSM_hook;
+ParallelWorkerMain_hook_type ParallelWorkerMain_hook;
```
Comment (Deepesh125) [inline src/backend/access/transam/parallel.c]: shouldn't we initialise it to NULL?

File: src/backend/access/transam/parallel.c [hunk 5]
```diff
@@ -465,7 +473,12 @@ InitializeParallelDSM(ParallelContext *pcxt)
 		strcpy(entrypointstate, pcxt->library_name);
 		strcpy(entrypointstate + lnamelen + 1, pcxt->function_name);
 		shm_toc_insert(pcxt->toc, PARALLEL_KEY_ENTRYPOINT, entrypointstate);
+
+		/* Initialize babelfish fixed-size state in shared memory. */
+		if (MyProcPort->is_tds_conn && InitializeParallelDSM_hook)
+			(*InitializeParallelDSM_hook) (pcxt, false);
 	}
+
```
Comment (Deepesh125) [inline src/backend/access/transam/parallel.c]: Un-necessary changes.

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: src/backend/access/transam/parallel.c (hunk 4 above)
Comment (ahmed-shameem) [reply src/backend/access/transam/parallel.c]: Updated in latest rev.

File: src/backend/access/transam/parallel.c (hunk 5 above)
Comment (ahmed-shameem) [reply src/backend/access/transam/parallel.c]: Updated in latest rev.

**Review by Deepesh125 — APPROVED** [review]

**Review by surendravishnoi — APPROVED** [review]
